### PR TITLE
Re-increment uncaught exception count on rethrow

### LINKED ISSUE #11997: Bug in std::uncaught_exceptions()
```
#include <iostream>
#include <exception>

int main()
{
    try
    {
        struct print { ~print() { std::cout << std::uncaught_exceptions() << std::endl; } } p;
        try
        {
            throw std::exception();
        }
        catch(...)
        {
            throw;
            // throw std::exception(); // <-- works correctly
        }
    }
    catch(...)
    {
    }
    return 0;
}
```

The program above prints `0` under emscripten. It should print `1`.

This appears to be due to incorrect handling of `throw;` -- the program works correctly if that statement is changed to throw a new exception object.

I'm using `emcc (Emscripten gcc/clang-like replacement) 2.0.0 (d81f400831a22f04901ab149f29d3caf77979bf6)` (of course with exception catching enabled).

### PR DESCRIPTION
**Pull request by firemansamm**
Fixes #11997, I think?

In https://github.com/emscripten-core/emscripten/blob/master/src/library_exceptions.js, the uncaught exception count is incremented and decremented exactly once during `__cxa_throw` and `__cxa_begin_catch` respectively (as the ABI says it should be). This however doesn't get re-incremented in the event of a rethrow, so anything that happens between the first catch and the second catch (eg. destructors) receive the decremented value.

In LLVM the count is incremented again in `__cxa_rethrow` (which curiously the ABI does not ask us to do, but seems to be the behaviour we expect).

I'm new around here, please don't roast me too hard if I'm way off the mark :)

### CODE REVIEW

**Comment by kripken** [comment]
Thanks @firemansamm !

cc @vagran who worked most recently on this and may remember best the ABI details.

> as the ABI says it should be

(it's been a while since I read those ABI details - do you have a link to that?)

**Comment by firemansamm** [comment]
@kripken, the reference I used when checking this behaviour was this one: https://refspecs.linuxfoundation.org/abi-eh-1.22.html

**Review by kripken — APPROVED** [review]
Thanks @firemansamm !

Reading the ABI doc, I am puzzled it doesn't say anything about rethrow incrementing the count, when it seems like it should, and like you say LLVM does. Maybe the ABI doc's assumption is that rethrow calls throw under the hood..?

Regardless, this patch looks right to me, and I verified the testcase has the right output with it (and is incorrect without it). I can't think of a better way to fix this testcase given that expected behavior.

I'm not an expert on these ABI details, so maybe @vagran or @sbc100 also want to take a look?

File: tests/core/test_exceptions_uncaught_count.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    try {
+      throw std::exception();
+    } catch (...) {
+      printf("in inner catch: %d\n", std::uncaught_exceptions());
+      throw;
+    }
+  } catch (...) {
+    printf("in outer catch: %d\n", std::uncaught_exceptions());
+  }
+}
```
Comment (sbc100) [inline tests/core/test_exceptions_uncaught_count.cpp:30]: missing newline

**Review by sbc100 — APPROVED** [review]
I'm also not an expert on the ABI details, but in general this change lgtm

**Comment by kripken** [comment]
@firemansamm I think those CI errors can be fixed by merging in latest trunk to here.

File: tests/core/test_exceptions_uncaught_count.cpp (hunk 1 above)
Comment (firemansamm) [reply tests/core/test_exceptions_uncaught_count.cpp]: Done :)

**Comment by firemansamm** [comment]
Thanks @kripken and @sbc100! I briefly messed up the merge, but it should be fine now :)

**Comment by kripken** [comment]
Perfect, thanks @firemansamm !

**Pull request merged**